tests/test_uncrustify/test.py

```diff
@@ -0,0 +1,211 @@
+                 self.test_name, self.test_name.split(':')[0]))
+
+    # -------------------------------------------------------------------------
+    def _diff(self, expected, actual):
+        sys.stdout.flush()
+        cmd = [config.git_exe, 'diff', '--no-index', expected, actual]
+        subprocess.call(cmd)
+
+    # -------------------------------------------------------------------------
+    def _run_pass(self, i, args):
+        _config = getattr(self, self.pass_config[i])
+        _input = getattr(self, self.pass_input[i])
+        _expected = getattr(self, self.pass_expected[i])
+        _result = os.path.join(args.result_dir, self.pass_result_dir[i],
+                               os.path.basename(os.path.dirname(_expected)),
+                               os.path.basename(_expected))
+
+        if args.verbose:
+            print(self.test_name + self.pass_note[i])
+            print('  Language : {}'.format(self.test_lang))
+            print('     Input : {}'.format(_input))
+            print('    Config : {}'.format(_config))
+            print('  Expected : {}'.format(_expected))
+            print('    Result : {}'.format(_result))
+
+        if not os.path.exists(self.pass_result_dir[i]):
+            os.makedirs(self.pass_result_dir[i])
+
+        cmd = [
+            config.uncrustify_exe,
+            '-q',
+            '-l', self.test_lang,
+            '-c', _config,
+            '-f', _input,
+            '-o', _result
+        ]
+        if args.debug:
+            cmd += [
+                '-LA',
+                '-p', _result + '.unc'
+            ]
+            stderr = open(_result + '.log')
+
+        else:
+            cmd += ['-L1,2']
+            stderr = None
+
+        if args.show_commands:
+            printc('RUN: ', repr(cmd))
+
+        try:
+            subprocess.check_call(cmd, stderr=stderr)
+        except subprocess.CalledProcessError as exc:
+            msg = '{}{}: Uncrustify error code {}'
+            msg = msg.format(self.test_name, self.pass_note[i], exc.returncode)
+            printc('FAILED: ', msg, **FAIL_ATTRS)
+            raise ExecutionFailure(exc)
+        finally:
+            del stderr
+
+        try:
+            if not filecmp.cmp(_expected, _result):
+                printc('{}: '.format(self.pass_diff_text[i]),
+                       self.test_name + self.pass_note[i],
+                       **self.pass_diff_attrs[i])
+                if args.diff:
+                    self._diff(_expected, _result)
+                raise self.pass_exception[i](_expected, _result)
+        except OSError as exc:
+            printc('MISSING: ', self.test_name + self.pass_note[i],
+                   **self.pass_diff_attrs[i])
+            raise MissingFailure(exc, _expected)
+
+    # -------------------------------------------------------------------------
+    def run(self, args):
+        self.check()
+
+        self._run_pass(0, args)
+        self._run_pass(1, args)
```

Comment:
As a note, I think the concept of a rerun and instability are two different things to test.

Reply:
...don't think I follow?

Reply:
A rerun config is a different config file (usually inheriting the first one and flipping some options around) whereas an instability is proven when it uses the same config but the results are different every time you run it.

Reply:
...and this makes sense if you think about using the rerun expected output as its own input; the point of a separate `.rerun` is to twiddle with the test so that it *will* be stable when it otherwise wouldn't be. IOW, the second pass is always a pure stability pass, but it may not use the same inputs as the "main" pass.

Reply:
On the other hand, I have seen tests which are not "stable" not even after 2 passes. I'm unsure what the purpose of a rerun config is then but it could be what you say: to have a passing when we know we have an instability. I'd prefer having this and turning instabilities as errors then silently allowing instabilities to creep in the codebase.

Reply:
So... that's what we have now? IOW, the last comment is just an agreement?

Note: Instabilities are errors (d66654cbf4b8). Separate re-run files are a way to skirt that, which can be "correct" (c:00012) or dodgy (#1907, #1908, #1909). These will never appear "silently", since tests have to be annotated (`!`) to turn on the separate re-run functionality, as well as having a re-run expected and/or config. Otherwise, an instability is (and has been, since CTest) a CI failure.